kotlin_engine/src/main/java/com/github/paylike/kotlin_engine/viewmodel/PaylikeEngine.kt

```diff
@@ -0,0 +1,287 @@
+    private fun checkValidState(validState: EngineState, callerFun: String) {
+        if (currentState != validState) {
+            throw InvalidEngineStateException(
+                "Can't call $callerFun in this state: $currentState\nThe valid state now is $validState"
+            )
+        }
+    }
+
+    /** Internal function to execute api call respecting [ApiMode] state */
+    private suspend fun payment(): PaylikeClientResponse {
+        if (repository.paymentRepository == null) {
+            throw InvalidPaymentDataException(
+                "PaymentData is invalid.\n${repository.paymentRepository!!::class.simpleName} is missing."
+            )
+        }
+        if (
+            repository.paymentRepository?.integration == null ||
+                repository.paymentRepository?.card == null
+        ) {
+            throw InvalidPaymentDataException("PaymentBody is not valid.")
+        }
+        if (this.apiMode == ApiMode.TEST && repository.paymentRepository?.test == null) {
+            throw InvalidPaymentDataException("PaymentBody is not valid. Test DTO is missing.")
+        }
+        val response: PaylikeClientResponse
+        coroutineScope {
+            response =
+                when (apiMode) {
+                    ApiMode.LIVE -> {
+                        apiService.paymentCreate(repository.paymentRepository!!)
+                    }
+                    ApiMode.TEST -> {
+                        apiService.paymentCreate(repository.paymentRepository!!)
+                    }
+                }
+        }
+        return response
+    }
+
+    /** Sets error corresponding to the cause */
+    private fun setErrorState(e: Exception) {
+        when (e::class.superclasses.first()) {
+            PaylikeException::class -> {
+                e as PaylikeException
+                log.accept("An API exception occurred: ${e.code} ${e.cause}")
+                error =
+                    PaylikeEngineError(
+                        e.message ?: "No exception message is included.",
+                        paylikeException = e
+                    )
+            }
+            EngineException::class -> {
+                e as EngineException
+                log.accept("An engine exception occurred: ${e.message}")
+                error =
+                    PaylikeEngineError(
+                        e.message ?: "No exception message is included.",
+                        engineException = e
+                    )
+            }
+            else -> {
+                log.accept("A not paylike nor engine exception has occurred: $e")
+                error = PaylikeEngineError(e.message ?: "No exception message is included.")
+            }
+        }
+        currentState = EngineState.ERROR
+    }
+
+    override fun notifyObservers() {
+        this.setChanged()
+        notifyObservers(currentState)
+    }
+
+    override fun notifyObservers(arg: Any?) {
```

Comment:
I don't think we need this right?

Reply:
This current solution was created to hide the arg when we call a nofityObservers, because the arg always must be engine state. Maybe it is overengineering so i changed it to only use the parameter version and it is explicitly called always.